### PR TITLE
issue/#56/rss-of-all-rsses

### PR DESCRIPTION
**Pull request by goostleek**

### CODE REVIEW

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/SpringApplicationContext.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package pl.tomaszdziurko.jvm_bloggers;
+
+import org.springframework.beans.BeansException;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SpringApplicationContext implements ApplicationContextAware {
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/SpringApplicationContext.java]: You can just Di ApplicationContext using @Autowired ...implementing interface is not needed..or you vae good reasons to do that ;]

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import pl.tomaszdziurko.jvm_bloggers.blog_posts.domain.BlogPost;
+import pl.tomaszdziurko.jvm_bloggers.blog_posts.domain.BlogPostRepository;
+import pl.tomaszdziurko.jvm_bloggers.utils.NowProvider;
+import pl.tomaszdziurko.jvm_bloggers.utils.UriUtmComponentsBuilder;
+
+@Service
+@CacheConfig(cacheNames = AggregatedRssFeedProducer.RSS_CACHE)
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: me gusta :)

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/SpringApplicationContext.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+    private static ApplicationContext context;
+
+    @Override
+    public void setApplicationContext(ApplicationContext context) throws BeansException {
+        SpringApplicationContext.context = context;
+    }
+
+    public static <T> T getBean(Class<T> clazz) {
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/SpringApplicationContext.java]: Why we need that?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    private final NowProvider nowProvider;
+
+    @Cacheable
+    public SyndFeed getRss() {
+
+        final StopWatch stopWatch = new StopWatch();
+        log.debug("Building agregated RSS feed...");
+        stopWatch.start();
+
+        final List<BlogPost> approvedPosts = blogPostRepository.findByApprovedTrueOrderByPublishedDateDesc();
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: Can we get null from Repository here?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        final List<SyndEntry> feedItems = approvedPosts.stream()
+                .map(this::toRssEntry)
+                .collect(Collectors.toList());
+
+        final SyndFeed feed = new SyndFeedImpl();
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: Builder here maybe?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        feed.setEntries(feedItems);
+
+        stopWatch.stop();
+        log.debug("Total {} feed entries produced in {}ms", feedItems.size(), stopWatch.getTotalTimeMillis());
+
+        return feed;
+    }
+
+    private SyndEntry toRssEntry(BlogPost post) {
+        final SyndEntry rssEntry = new SyndEntryImpl();
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: Builder here maybe?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        rssEntry.setTitle(post.getTitle());
+        rssEntry.setLink(addUtmComponents(post.getUrl()));
+        rssEntry.setAuthor(post.getBlog().getAuthor());
+        rssEntry.setPublishedDate(toDate(post.getPublishedDate()));
+
+        final String description = post.getDescription();
+        if (isNotBlank(description)) {
+            final SyndContentImpl descriptionContent = new SyndContentImpl();
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: Why not pass it to constructor? Immutability?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/BlogPostsController.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+
+@RestController
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class BlogPostsController {
+
+    private final AggregatedRssFeedProducer rssProducer;
+
+    @SneakyThrows
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/BlogPostsController.java]: Hmm...I do not think that "masking" exceptions is a good idea on REST endpoints;/ I am just thinking what Swagger2 will say when we decide to use it in the future? Can we just use @ControllerAdvice and handle exception?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPostEntityListener.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+    @PostPersist
+    @PostUpdate
+    @PostRemove
+    public void invalidateRssCache(BlogPost blogPost) {
+        getCacheManager().getCache(AggregatedRssFeedProducer.RSS_CACHE).clear();
+    }
+
+    private CacheManager getCacheManager() {
+        return SpringApplicationContext.getBean(CacheManager.class);
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPostEntityListener.java]: Maybe we could use @Configurable: http://docs.spring.io/autorepo/docs/spring-framework/3.2.16.RELEASE/javadoc-api/org/springframework/context/annotation/aspectj/EnableSpringConfigured.html

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/utils/NowProvider.java [hunk 10]
```diff
@@ -9,9 +9,9 @@
 @Component
 public class NowProvider {
 
-    public static final String ZONE_NAME = "Europe/Paris";
+    public static final ZoneId DEFAULT_ZONE = ZoneId.of("Europe/Warsaw");
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/utils/NowProvider.java]: Why not Europe/Cracow? :D

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java [hunk 11]
```diff
@@ -33,4 +34,8 @@ public Client getMailingRestClient(@Value("${mailing.apiKey}") String malingApiK
         return client;
     }
 
+    @Bean
+    public CacheManager cacheManager() {
+        return new GuavaCacheManager();
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: Maybe simple ConcurrentHashMap would be enough?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+    private static final String UTM_MEDIUM = "RSS";
+
+    private final BlogPostRepository blogPostRepository;
+    private final NowProvider nowProvider;
+
+    @Cacheable
+    public SyndFeed getRss() {
+
+        final StopWatch stopWatch = new StopWatch();
```
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: We could use @Around advice to to profile methos invocations, what do you think?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java (hunk 11 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: With ConcurrentHashMap and own implementation you are writing your own cache instead of using already existing one :)

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/SpringApplicationContext.java (hunk 3 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/SpringApplicationContext.java]: I think it is to get use this bean in a `BlogPostEntityListener`, which is created and executed not in the Spring managed scope.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        log.debug("Building agregated RSS feed...");
```
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: agregated -> aggregated

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+        final SyndFeed feed = new SyndFeedImpl();
+        feed.setFeedType(FEED_TYPE);
+        feed.setTitle(FEED_TITLE);
+        feed.setDescription(FEED_DESCRIPTION);
+        feed.setPublishedDate(toDate(nowProvider.now()));
+        feed.setEntries(feedItems);
+
+        stopWatch.stop();
+        log.debug("Total {} feed entries produced in {}ms", feedItems.size(), stopWatch.getTotalTimeMillis());
```
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: Maybe info here? This should not add too much noise to the logs and can be useful to see how often and how log this process takes?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java (hunk 6 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: This is not our class, it is from Rome :)

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java (hunk 7 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: Same :) This is class from Rome library, with only no-args constructor.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+            final SyndContentImpl descriptionContent = new SyndContentImpl();
+            descriptionContent.setValue(description);
+            rssEntry.setDescription(descriptionContent);
+        }
+
+        return rssEntry;
+    }
+
+    private String addUtmComponents(String url) {
+        final String campaign = String.format("RSS@%s",
```
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: Not sure we want to have different campaign each time when RSS is regenerated. Maybe it should be constant for RSS feed?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java (hunk 11 above)
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: Yes...but as I can see this one is used right now only to store/retrive data...for me: **overkill ** at this very moment.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/BlogPostsController.java (hunk 8 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/BlogPostsController.java]: Hmm... I think for now it's good enough. When we introduce some kind of monitoring or metrics, we could think about adding @ControllerAdvice to get all exceptions in one place.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/NewBlogPostStoringActor.java [hunk 16]
```diff
@@ -39,7 +39,7 @@ private BlogPost createBlogPost(RssEntryWithAuthor rssEntry) {
         return BlogPost.builder()
                 .title(postInRss.getTitle())
                 .url(postInRss.getLink())
-                .publishedDate(DateTimeUtilities.convertDateToLocalDateTime(dateToStore))
+                .publishedDate(DateTimeUtilities.toLocalDateTime(dateToStore))
```
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/NewBlogPostStoringActor.java]: Much better method name :+1:

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPostEntityListener.java (hunk 9 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPostEntityListener.java]: @adamdec I have never tried it but it can be a nice addition/change. Please feel free to introduce it in a separate PR :)

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java (hunk 4 above)
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: No.

**Comment by tdziurko** [comment]
Review done, waiting for some tiny fixes :)

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java (hunk 11 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: Hmm.. we are not on the same page here :) Cache is mostly used to store/retrieve data. Not sure what are you referring to :)

Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: I am refering to fact that in this case we are not using anything "special" from Guava. SO if this is simple get/put...why not simple ConcurrentMap...KISS :)

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java (hunk 5 above)
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: It's class from _Rome_ and it has only one constructor.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java (hunk 11 above)
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: @adamdec: Reinventing the wheel is not KISS IMHO. Why writing custom cache implementation based on `ConcurrentHashMap` if we can use Spring provided [Cache Abstraction](https://docs.spring.io/spring/docs/current/spring-framework-reference/html/cache.html) with Guava cache as implementation provider? We don't even have to add any additional dependencies, we already have it in the project. Where do you see overkill here? IMHO using Spring Cache Abstraction has many advantages over custom cache implementation.:
- writing custom cache (even simple map/get/put approach) involves writing custom code and every additional custom line of code increases probability of bugs and errors (opposition of KISS)
- _Spring Cache Abstraction_ can be used in a clean and declarative way, just add @Cacheable annotation and you got caching (in opposite to custom based implementation)

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java (hunk 12 above)
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/AggregatedRssFeedProducer.java]: It's rather to put log info rather than real profiling.

**Comment by goostleek** [comment]
All requested review changes done.

**Comment by goostleek** [comment]
All requested review changes done.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java (hunk 11 above)
Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: @goostleek Chill :) What I meant by KISS in the example of yours is to use "The JDK-based Cache implementation resides under org.springframework.cache.concurrent package. It allows one to use ConcurrentHashMap as a backing Cache store." Please see: 35.6.1 JDK ConcurrentMap-based Cache

- You do not have to.
- Yes, this is what we are doing right now

I think there is a bit of misunderstanding here :)

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPostEntityListener.java (hunk 9 above)
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPostEntityListener.java]: I'm afraid it's not easily possible at the moment: https://github.com/spring-projects/spring-boot/issues/739

**TLDR;**
Having _Spring Boot_ application packaged into a single distributable uber jar will require distributing it with additional externally provided dependency (aspectj.jar) and setting load-time weaving by adding `-javaagent` path to AspectJ weaving agent to startup script in order to get `@Configurable` annotation working. It will also complicate development and testing environment as it will require additional development environment setup.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java (hunk 11 above)
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: Ok , I understand you concern now. You are talking about using default [`CacheManager`](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/cache/concurrent/ConcurrentMapCacheManager.html) provided by Spring Boot OotB, right?

Comment (adamdec) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/JvmBloggersConfiguration.java]: Finally!! :)
